Assets/Scripts/StandardSamples/UI/Achievements/UIAchievementsMenu.cs

```diff
@@ -201,33 +147,10 @@ private async void OnAchievementDataUpdated()
                 scrollRect.gameObject.SetActive(true);
                 scrollRect.content.sizeDelta = new Vector2(0, achievementDefCount * 30);
 
-                int i = 0;
                 List<AchievementData> achievementDataListCopy = new(achievementDataList);
                 foreach (var achievementData in achievementDataListCopy)
                 {
-                    var newButton = Instantiate(itemTemplate, achievementListContainer);
-                    newButton.SetNameText(achievementData.Definition.AchievementId);
-                    newButton.gameObject.SetActive(true);
-                    newButton.index = i;
-                    bool unlocked = achievementData.PlayerData.HasValue && achievementData.PlayerData.Value.Progress >= 1;
-                    Texture2D iconTex = null;
-                    int iconGiveupFrame = Time.frameCount + 120;
-                    while (iconTex == null)
-                    {
-                        iconTex = unlocked ?
-                            AchievementsService.Instance.GetAchievementUnlockedIconTexture(achievementData.Definition.AchievementId)
-                           : AchievementsService.Instance.GetAchievementLockedIconTexture(achievementData.Definition.AchievementId);
-                        await System.Threading.Tasks.Task.Yield();
-
-                        if (Time.frameCount > iconGiveupFrame)
-                        {
-                            UnityEngine.Debug.LogWarning("Timeout : Failed to get icon");
-                            break;
-                        }
-                    }
-                    newButton.SetIconTexture(iconTex);
-                    i += 1;
-                    achievementListItems.Add(newButton);
+                    await AddAchievementButton(achievementData);
```

Comment:
Could this possibly be set up to handle getting all of these at the same time / await all, instead of awaiting one at a time?

Reply:
The reason it does them one at a time is because the downloading of the image bytes happens one at a time. That is - we don't want to implement it in such a way that it requires all the images to be loaded. This implementation lets them load asynchronously.